lib/view_component_reflex/state_adapter/redis.rb

```diff
@@ -0,0 +1,74 @@
+# A redis based adapter for component_reflex
+#
+# ViewComponentReflex::Engine.configure do |config|
+#   config.state_adapter = ViewComponentReflex::StateAdapter::Redis.new(
+#       redis_opts: {
+#           url: "redis://localhost:6379/1", driver: :hiredis
+#       },
+#       ttl: 3600)
+# end
+
+module ViewComponentReflex
+  module StateAdapter
+    class Redis
+      attr_reader :client, :ttl
+
+      def initialize(redis_opts:, ttl: 3600)
+        @client = Redis.new(redis_opts)
+        @ttl = ttl
+      end
+
+      def state(request, key)
+        cache_key = get_key(request, key)
+        value = client.hgetall(cache_key)
+
+        return {} if value.nil?
+
+        value.map do |k, v|
+          [k, JSON.parse(v)]
+        end.to_h
+      end
+
+      def set_state(request, _, key, new_state)
+        cache_key = get_key(request, key)
+        save_to_redis(cache_key, new_state)
+      end
+
+      def store_state(request, key, new_state = {})
+        cache_key = get_key(request, key)
+        optimized_store_to_redis(cache_key, new_state, request)
+      end
+
+      def wrap_write_async
+        client.pipelined do
+          yield
+        end
+      end
+
+      private
+
+      # Reduce number of calls coming from #store_state to save Redis
+      # when it's first rendered
+      def optimized_store_to_redis(cache_key, new_state, request)
+        request.env["CACHE_REDIS_REFLEX"] ||= []
+        return if request.env["CACHE_REDIS_REFLEX"].include?(cache_key)
+        request.env["CACHE_REDIS_REFLEX"].push(cache_key)
+
+        save_to_redis(cache_key, new_state)
+      end
+
+      def save_to_redis(cache_key, new_state)
+        new_state_json = new_state.map do |k, v|
+          [k, v.to_json]
```

Comment:
This to_json call and the corresponding parse would prevent a user from storing a model in state, and operating on it as if it was an instance variable. Instead, they would be operating on a hash.

The todo component is a good example in the expo: https://github.com/joshleblanc/view_component_reflex_expo/blob/master/app/components/todo_component.rb

What we could do instead is dump the object, which is what the session does by default.

```suggestion
          [k, Marshal.dump(v)]
```